python/nav/portadmin/napalm/juniper.py

```diff
@@ -0,0 +1,218 @@
+#
+# Copyright (C) 2020 Uninett AS
+#
+# This file is part of Network Administration Visualized (NAV).
+#
+# NAV is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License version 3 as published by
+# the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+# details.  You should have received a copy of the GNU General Public License
+# along with NAV. If not, see <http://www.gnu.org/licenses/>.
+#
+"""Juniper specific PortAdmin functionality.
+
+Some references:
+- https://www.juniper.net/documentation/en_US/junos-pyez/information-products/pathway-pages/junos-pyez-developer-guide.html
+- https://www.juniper.net/documentation/en_US/junos-pyez/topics/reference/general/junos-pyez-tables-op-predefined.html
+
+NAPALM is used as the base framework for fetching and setting configuration here,
+but many of the operations PortAdmin needs are not directly supported by the NAPALM API,
+so the underlying Juniper PyEZ library is utilized directly in most cases.
+
+"""
+from operator import attrgetter
+from typing import List, Any, Dict, Tuple
+
+import napalm
+from napalm.base.exceptions import ConnectAuthError, ConnectionException
+
+from nav.enterprise.ids import VENDOR_ID_JUNIPER_NETWORKS_INC
+from nav.models import manage
+from nav.portadmin.handlers import (
+    ManagementHandler,
+    DeviceNotConfigurableError,
+    AuthenticationError,
+    NoResponseError,
+)
+from nav.junos.nav_views import (
+    EthPortTable,
+    EthernetSwitchingInterfaceTable,
+    InterfaceConfigTable,
+)
+from jnpr.junos.op.vlan import VlanTable
+
+from nav.portadmin.vlan import FantasyVlan
+from nav.util import first_true
+
+__all__ = ["Juniper"]
+
+# This maps interface oper/admin status values to SNMP values as used in NAV's data
+# model. See IF-MIB::ifOperStatus and IF-MIB::ifAdminStatus from RFC 2863 for details.
+SNMP_STATUS_MAP = {"up": 1, "down": 2, True: 1, False: 2}
+
+
+class Juniper(ManagementHandler):
+    """Juniper specific version of a Napalm PortAdmin handler.
+
+    Juniper switches do things a bit differently from the standard model supported by
+    NAV. VLAN config is put on logical sub-units of the physical ports. SNMP-wise,
+    these count as separate, but related interfaces.
+
+    NAV's standard interpretation from collection is that the logical sub-units are
+    switch ports, since these are the interfaces listed in the BRIDGE-MIB, while the
+    parent ports are largely ignored as significant.
```

Comment:
I did. But I see this description might need to be updated as well - because this fact also varies between ELS switches and non-ELS switches, in that ELS switches seem to report the VLAN data on the physical interfaces on SNMP, even though the configuration still belongs on the sub-units.